rust/foxglove/src/websocket/ws_protocol/client/player_state.rs

```diff
@@ -0,0 +1,201 @@
+#[cfg(feature = "unstable")]
+use bytes::{Buf, BufMut};
+
+#[cfg(feature = "unstable")]
+use crate::websocket::ws_protocol::{BinaryMessage, ParseError};
+
+#[cfg(feature = "unstable")]
+use super::BinaryOpcode;
+
+#[cfg(feature = "unstable")]
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum PlaybackState {
+    Playing = 0,
+    Paused = 1,
+    Buffering = 2,
+    Ended = 3,
+}
+
+#[cfg(feature = "unstable")]
+impl TryFrom<u8> for PlaybackState {
+    type Error = u8;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::Playing),
+            1 => Ok(Self::Paused),
+            2 => Ok(Self::Buffering),
+            3 => Ok(Self::Ended),
+            _ => Err(value),
+        }
+    }
+}
+
+#[cfg(feature = "unstable")]
+#[doc(hidden)]
+/// Player state message.
+#[derive(Debug, Clone, PartialEq)]
+pub struct PlayerState {
+    /// Playback state
+    pub playback_state: PlaybackState,
+    /// Playback speed
+    pub playback_speed: f32,
+    /// Seek playback time in nanoseconds (only set if a seek has been performed)
+    pub seek_time: Option<u64>,
+}
+
+#[cfg(feature = "unstable")]
+impl<'a> BinaryMessage<'a> for PlayerState {
+    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
+        // Message size: playback_state (1 byte) + playback_speed (4 bytes) + had_seek (1 byte) + seek_time (8 bytes)
+        if data.len() < 1 + 4 + 1 + 8 {
+            return Err(ParseError::BufferTooShort);
+        }
+
+        let state_byte = data.get_u8();
+        let playback_state = PlaybackState::try_from(state_byte)
+            .map_err(|_| ParseError::InvalidPlaybackState(state_byte))?;
+
+        let playback_speed = data.get_f32_le();
+        let had_seek = data.get_u8() != 0;
+        let seek_time = if had_seek && data.len() >= 8 {
```

Comment:
Since L53 validates length:

```suggestion
        let seek_time = if had_seek {
```